pkg/codegen/go/gen.go

```diff
@@ -1638,6 +1638,13 @@ func (pkg *pkgContext) genResource(w io.Writer, r *schema.Resource, generateReso
 		fmt.Fprint(w, "\topts = append(opts, replaceOnChanges)\n")
 	}
 
+	// Setup PluginDownloadURL. We pre-append our call to PluginDownloadURL to
+	// that if a user adds their own separate PluginDownloadURL, we won't
+	// override it.
+	if url := pkg.pkg.PluginDownloadURL; url != "" {
+		fmt.Fprintf(w, "\topts = append([]pulumi.ResourceOption{pulumi.PluginDownloadURL(%q)},  opts...)\n", url)
```

Comment:
@pgavlin, quick question that's orthogonal to this PR, but kind of related:

Shouldn't we be setting the default version in here like we do in the other languages, or am I missing something?

https://github.com/pulumi/pulumi/blob/684d7aad3920042ce1982b5fedddea62a6e5bb61/pkg/codegen/dotnet/gen.go#L1037

https://github.com/pulumi/pulumi/blob/684d7aad3920042ce1982b5fedddea62a6e5bb61/pkg/codegen/nodejs/gen.go#L868-L870

https://github.com/pulumi/pulumi/blob/684d7aad3920042ce1982b5fedddea62a6e5bb61/pkg/codegen/python/gen.go#L1219-L1220

Maybe we didn't do this for Go because the version info for Go isn't readily available? (Although, we did later add [`PkgVersion`](https://github.com/pulumi/pulumi/blob/684d7aad3920042ce1982b5fedddea62a6e5bb61/pkg/codegen/go/gen.go#L3610-L3623) used in resource module registrations; so perhaps we could consider using that, if it was desirable).

Reply:
Honestly, I'm not sure why we didn't do this in Go. It does seem like we should. I'm also not sure how we arrived at `PkgVersion`--any idea if there's a reason we didn't just embed the version statically? The reflection-based code used by `PkgVersion` is essentially static, I think.

Reply:
We can't embed the version statically because we can't trust that the version supplied in the schema is correct. This is the same reason why https://github.com/pulumi/pulumi/issues/8521 exists, and is blocked. Injecting the version at codegen time would require changing how our provider makefile work.

Reply:
Ah yep, thank you for the reminder. I wonder if we should switch this to using a `//go:embed` and a parse in `init` so that we can get a real version. This would be akin to the solutions we use for the other language. More concretely, we could e.g. have the Go code generator emit a dummy `version.txt` file that is embedded into the package, then have the Makefile update the contents of that file with the actual server. The contents of the file would be parsed into a real version during the package's `init` function.

Reply:
If we are willing to touch every Makefile to actualize that change, I think we should instead attempt to find a safe way to embed version into the binary (if at all possible).

Reply:
> I think we should instead attempt to find a safe way to embed version into the binary (if at all possible).

I would love it if we could find a way to do this, but I'll admit that I'm pessimistic about making that happen any time soon. I think that taking a tactical fix--even though it requires updating Makefiles--is a chance for us to not let perfect be the enemy of better.

Reply:
> we could e.g. have the Go code generator emit a dummy `version.txt` file that is embedded into the package, then have the Makefile update the contents of that file with the actual server

I still don't quite follow how this would work for go modules. Say we generate an SDK for a `foo` provider, emitted at `github.com/pulumi/pulumi-foo/sdk/go/foo`. Versions are tagged with git tags (e.g. `v1.2.3` and `sdk/v1.2.3`). Wouldn't we have to commit the `version.txt` file before tagging it? I don't see how the `Makefile` helps. When I use the `foo` module from my program, our `Makefile` is not involved.

Reply:
That's a good point. For Go, it's the push to the git remote that's the publishing step, so even if we took my suggestion we'd still be emitting the wrong version.